### ISSUE TITLE
Server Side Rendering: First paint missing some styles (FOUC)

### ISSUE THREAD

**Issue by Nemsae** [styled-components/styled-components#1860]
## System:
 - OS: macOS High Sierra 10.13.5
 - CPU: x64 Intel(R) Core(TM) i5-4308U CPU @ 2.80GHz
 - Memory: 82.71 MB / 8.00 GB
 - Shell: 3.2.57 - /bin/bash
## Binaries:
 - Node: 8.9.4 - ~/.nvm/versions/node/v8.9.4/bin/node
 - npm: 6.1.0 - ~/temp/Martin/metrics-dashboard/node_modules/.bin/npm
 - Watchman: 4.7.0 - /usr/local/bin/watchman
## npmPackages:
 - babel-plugin-styled-components: ^1.5.1 => 1.5.1 
 - styled-components: ^3.2.6 => 3.2.6 

## Reproduction

The chunk of server side rendering code used:
```
        const sheet = new ServerStyleSheet()
        let html = ReactDOMServer.renderToString(sheet.collectStyles(
          <Provider store={store}>
            {<RouterContext {...renderProps} />}
          </Provider>
        ))
        const styleTags = sheet.getStyleTags() // or sheet.getStyleElement()
```
 My index template (index.ejs) as such:
```
<!DOCTYPE html>
<html>

<head>
  <%- styleTags %>
</head>
<body style="margin: 0; font-family: 'PT Sans', Arial;">
  <div id="root" style="height: 100%;">
    <%- html %>
  </div>
  <script type="text/javascript" charset="utf-8">
    window.__REDUX_STATE__ = '<%= reduxState %>';
  </script>

</body>
</html>
```
And just in case some styled-component code, though I don't expect this to be the issue (possibly because of using `extend`?).
```
//   H1.js file

import styled from 'styled-components';

const H1 = styled.h1.attrs({ className: 'h1' })`
    font-size: 4rem;
`;

export default H1;

//   end of H1.js file

//   WidgetTitle.js file
import NormalH1 from "components/H1";

const WidgetTitle = NormalH1.extend.attrs({ className: 'widget__title' })``;

export default WidgetTitle;

//   end of WidgetTitle.js file
```

## Steps to reproduce
1. Server side rendering set up as shown above, following closely the steps here: https://www.styled-components.com/docs/advanced#server-side-rendering.

## Expected Behavior
For all styles to be applied during first paint (like so below).

![First paint expected](https://user-images.githubusercontent.com/20247504/43098949-9ab107be-8e8e-11e8-8df6-95a5bdf48c11.png)

## Actual Behavior
Only some styles being applied on first paint, while some rules (specifically `font-size`) **seems** to be applied only after the JS bundle is parsed and executed. 

![first paint missing font size rules](https://user-images.githubusercontent.com/20247504/43098898-750126fc-8e8e-11e8-89eb-39aa71d09594.png)

## Remarks
As you can see in the two screenshots, I get a flash of content where the `font-size` rule isn't respected. I had this same issue with `font-family` actually, but dealt with it by putting it as an inline style on my `<body>` tag. Possibly this is a Flash of Unstyled Text (FOUT), but I'm honestly quite lost as to why these specific rules are being ignored on the first paint. 

I state that the rules **seem** to be applied only after the main JS bundle is loaded/parsed/executed because of examining the request for `app.js` closely in the Chrome Network panel. Is this possible a webpack issue where I need to optimize the bundle further?

**Comment by probablyup** [comment]
Out of curiosity, do you see the same issue with this code instead?

```js
const WidgetTitle = styled(NormalH1).attrs({ className: 'widget__title' })``;
```

**Comment by Nemsae** [comment]
@probablyup thanks for the suggestion - the issue still persists.

**Comment by probablyup** [comment]
Hmm I'm curious what it could be. Zocdoc has a pretty robust SSR setup with styled components and many types of components in use, haven't seen this sort of thing.

Any chance you could put together a lightweight reproduction, ideally in a codesandbox?

**Comment by wmertens** [comment]
Can you compare the rules you get in html with the rules you get after first render? Should be helpful to know exactly which rules are missing.

Ideally, you don't run data fetching, so client and server should be identical.

**Comment by Nemsae** [comment]
@wmertens Just to be clear you want to me compare the rules from the style tags (with the long string of sc-classes)
![so_sc-fout_1](https://user-images.githubusercontent.com/20247504/43550566-000e3b20-95b2-11e8-91a8-88c17479066b.png)

with the rules from the style tags received after first render, 
![so_sc-fout_2](https://user-images.githubusercontent.com/20247504/43550573-0753ee66-95b2-11e8-877b-b7c1df7d97a4.png)

The **FOUT** does disappear after receiving the second set of styles, after first render.

I've taken a quick look at the content of both sets of the initial style tags and immediately noticed that the styles are clipped, the `disp...` (unless chrome is clipping the content?).

![so_sc-fout_3](https://user-images.githubusercontent.com/20247504/43550807-cafc773e-95b2-11e8-9459-738de067a62e.png)

**Comment by Nemsae** [comment]
@probablyup I'll work on getting a sandbox up soon, hopefully before next week.

**Comment by wmertens** [comment]
Yes, correct. You can read the first set in your HTML file under the
sources tab, and the second you can double-click in Elements to see the
entire contents. Probably best to copy to an editor for easy comparison.
I'm suspecting some parsing or minifying issue. Did you try turning off the
Babel plugin minification/preprocessing?

**Comment by Nemsae** [comment]
I've compared the first set of tags (from the **sources** tab) with the second set of tags (from the **elements** tab). 

### Sources Style Tags
**Sources** only has one style tag that corresponds to the first style tag from the **elements** style tag set. 

Here's a diff check of the two files. [Diff check of sc-source.css and sc-elements-1.css](https://www.diffnow.com/?report=o7542)

### Elements Style Tags
**Elements** has three style tags, the first one which is part of the diff check from above. The second and third are the ones missing from **sources**. Notably, the third one contains my `injectGlobal` styles. 

I've attached a zip file containing the files for all four style tags. 
[styled-components-sources-and-elements-css-files.zip](https://github.com/styled-components/styled-components/files/2284328/styled-components-sources-and-elements-css-files.zip)

@wmertens I've checked my babel and webpack configuration for my `development` environment. No minification is being done. 

Here's my `.babelrc` file: 
```
{
  "presets": ["es2015", "stage-2", "react"],
  "env": {
    "test": {
      "plugins": [ "babel-plugin-rewire" ]
    },
    "development": {
      "plugins": [
        [
          "styled-components",
          {
            "ssr": true,
            "minify": false
          }
        ]
      ]
    },
    "production": {
      "plugins": [
        [
          "styled-components"
        ]
      ]
    }
  }
}

```

And here is my webpack configuration: 

```
var path = require('path')
var webpack = require('webpack')
var AssetsPlugin = require('assets-webpack-plugin')

var DEBUG = !(process.env.NODE_ENV === 'production')

if (DEBUG) {
  require('dotenv').config()
}

var config = {
  devtool: DEBUG ? 'cheap-module-eval-source-map' : false,
  entry: {
    app: ['./app/app'],
    vendor: [
      'react',
      'react-router',
      'redux',
      'react-dom',
      'lodash',
      'bluebird',
      'history'
    ]
  },
  resolve: {
    modules: [ path.join(__dirname, 'app'), 'node_modules' ]
  },
  output: {
    path: path.join(__dirname, 'dist'),
    filename: '[name].js'
  },
  plugins: [
    new webpack.EnvironmentPlugin(['NODE_ENV', 'API_BASE_URL'])
  ],
  module: {
    rules: [
      {
        test: /\.js$/,
        loader: 'babel-loader',
        exclude: /node_modules/,
        include: __dirname
      },
      {
        test: /\.(css|less)$/,
        use: [{
          loader: "style-loader" // creates style nodes from JS strings
        }, {
          loader: "css-loader" // translates CSS into CommonJS
        }, {
          loader: "less-loader" // compiles Less to CSS
        }]
      },
      {
          test: /\.(woff(2)?|ttf|eot|svg)(\?v=\d+\.\d+\.\d+)?$/,
          use: [{
              loader: 'file-loader',
              options: {
                  name: '[name].[ext]',
                  outputPath: 'fonts/'
              }
          }]
      },
      {
        test: /\.json$/,
        use: 'json-loader'
      }
    ]
  }
}


if (DEBUG) {
  config.entry.app.push('webpack-hot-middleware/client?path=/__webpack_hmr')

  config.plugins = config.plugins.concat([
    new webpack.HotModuleReplacementPlugin(),
    new webpack.optimize.CommonsChunkPlugin({
      name: 'vendor',
      filname: 'vendor.js'
    })
  ])
  config.output.publicPath = '/'
  config.module.rules.unshift({
    test: /\.js$/,
    loader: 'react-hot-loader',
    exclude: /node_modules/,
    include: __dirname
  })
} else {
  config.plugins = config.plugins.concat([
    new webpack.optimize.CommonsChunkPlugin({
      name: 'vendor',
      filname: '[name].[chunkhash].js'
    }),
    new webpack.optimize.UglifyJsPlugin(),
  ])
}

module.exports = config
```

Unless I've overlooked other possible minification/parsing settings, I suspect the issue arises from the second and third set of rules from **elements** not being supplied with the first trip back from SSR.

**Comment by probablyup** [comment]
Why aren’t you doing ssr: true  for production as well?

**Comment by mxstbr** [comment]
Yeah you need to do `ssr: true` for production. If you don't want the inferred display names, you can do `displayName: false` in production though!

**Comment by Nemsae** [comment]
I've only been attempting to debug this issue on my dev environment, expecting that both production and dev would replicate the same results. As per your instruction I've run the app in the production environment with the `ssr` flag on and still have the same issue. 

Also @probablyup if you could explain, my understanding from the docs is that the `ssr: true` option only avoids the checksum mismatch. How would a checksum mismatch prevent a portion of the style rules being delivered?

**Comment by probablyup** [comment]
Just trying to rule things out! Okay, another question... do you have React
code that’s like:

if(window) this
else that

Basically any code that happens during render that depends on server vs
client. It’s very important to render the same thing in both places and
then do whatever extra browser stuff you want in componentDidMount.

**Comment by StevenLangbroek** [comment]
I'm experiencing this same problem I think, except making a minimal repro is going to be a bit complicated due to the size of our codebase.

### How we do SSR
I imagine this is what most people are doing, but for each request we initialize our Redux store, get matching components for a route, trigger a static method on the class (which potentially kicks off some sagas), then when all sagas have completed we start actually rendering. This is basically gathering a renderStream into an accumulator and then `res.send`ing them over.

### Actual behavior
1. There are components that never get styled, and some that get styled to varying degrees. I'll focus on the footer, which never gets styled.
2. If I disable SSR (send empty string in root), everything works correctly.
3. Some of our pages don't have a footer. If I disable the footer on client, and then reenable it, styles get applied correctly

### Some further notes
- We only use `injectGlobal` for injecting font-face rules. It works correctly
- We don't use `.extend`
- I synchronized the babel plugin configuration & order between environments 100% while I'm working on this spike. `ssr` flag is `true` for all configurations, minimize etc have all been disabled.

**Comment by StevenLangbroek** [comment]
(I've tried both `3.4.2` and `3.2.3` btw, and am using v `1.5.1` of the babel plugin.

**Comment by MFCo** [comment]
I'm having the same issue.

I'm using NextJS, the following screen is in production environment (it's empty):
![image](https://user-images.githubusercontent.com/7811417/44101911-c4608394-9fe8-11e8-960e-8fa87336972f.png)

While the same code in development environment (if you expand it, it contains the styles):
![image](https://user-images.githubusercontent.com/7811417/44101997-fc257b36-9fe8-11e8-8a68-2590965d2380.png)

If I print the content of `sheet.getStyleElement()` bein sheet a ServerStyleSheet inside getInitialProps it contains the data I want to have in my tag, but it is not added.

**Comment by StevenLangbroek** [comment]
> That defeats the purpose of streaming.

@probablyup it defeats **a** purpose, but we still get the benefit of freeing up the event loop.

**Comment by MFCo** [comment]
@probablyup Thanks! But it doesn't explain why in smaller projects the tag is not empty. I'll continue looking for a solution to my issue.

It only loads the base style of each component, without media queries, but if I reload with the new component in the view port it loads correctly with the right media query but not other media queries or base style is applied at all.

**Comment by nandosangenetto** [comment]
I'm getting the same problem. 

I'm using the babel plugin with babel 7.
My injectGlobal is not being applied to our styles, resulting in FOUC (fonts and reset). Some doesn't get styles from SSR.

I don't know what I am missing.

**Comment by probablyup** [comment]
We might not be B7 compatible yet? https://github.com/styled-components/babel-plugin-styled-components/pull/134

**Comment by kitten** [comment]
Some things here we’ll need to understand what’s going on is:

- output of SSR (the html without JS modifications) which is easy to get with curl / httpie / devtools’ sources tabs
- code of SSR generator
- are you using unstable features like the stylesheetmanager during runtime?
- are you using streamed SSR?
- do you do anything unusual with StyledComponents global StyleSheet or it’s methods?

It might also be that several people are seeing several issues. One might be dealing with an invalid usage while another has some invalid CSS in their output.

So please bear with us and post some info please :) otherwise we’ll have to close this issue and ask everyone to create individual reproductions. Sorry 😅

**Comment by Nemsae** [comment]
@kitten Here's a [repl](https://repl.it/@jsson77/KindTintedWireframe) with both files of my SSR code and the html from the output of the SSR.

1. Not using StyleSheetManager.
2. Not using streamed SSR.
3. Not using any StyleSheet methods outside of my SSR file. 

Do let me know if you need anything else.

**Comment by wmertens** [comment]
Am I correct in thinking you only do one render pass? Don't you have
asynchronous data loading? Won't that change the mounted components?

I do two passes, and i wait for data to load in between. I only collect
styles for the second pass.

**Comment by Nemsae** [comment]
@wmertens I do have asynchronous data loading. 

But the data loading lifecyle happens after the **FOUC** disappears. I've taken some screenshots in Chrome Sources panel to illustrate this.

1. **FOUC** screen - which seems to happen while `app.js` bundle is being loaded/parsed.
![sc_load_1](https://user-images.githubusercontent.com/20247504/44630318-40294580-a929-11e8-998e-711c165b4202.png)

2. **NO FOUC** screen - all styles are now loaded, which happens after `app.js` is parsed.
![sc_load_2](https://user-images.githubusercontent.com/20247504/44630321-5505d900-a929-11e8-9592-b3c159783e78.png)

3. **DATA LOAD** screen - all styles already loaded, app now makes the `metric?security_k...` call to load the data. 
![sc_load_3](https://user-images.githubusercontent.com/20247504/44630323-5cc57d80-a929-11e8-834b-e464fb49607c.png)

Hopefully that answers your question. 

Could you help me understand how you are collecting the styles for the second pass?  Are you showing a loading component with inline styles first, then rendering the data bound components (with SC styles) once loading is done?

**Comment by probablyup** [comment]
@Nemsae could you check out 3.4.6 and let me know if it resolves the issue?

**Comment by StevenLangbroek** [comment]
@probablyup i'm upgrading now across our codebase, I'll merge my streaming spike back in and see if it fixes things.

**Comment by Nemsae** [comment]
This is my `app.js` with the `ReactDOM.render`: 

```
import "regenerator-runtime/runtime";
import 'babel-polyfill'

import React from 'react'
import ReactDOM from 'react-dom'
import { browserHistory } from 'react-router'

import configureStore from './configureStore'
import createRoutes from 'routes/index'
import { Provider } from 'react-redux'
import _ from 'lodash'

// Import CSS reset and Global Styles
import './global-styles'
import rootSaga from './rootSaga'

let reduxState = {}
if (window.__REDUX_STATE__) {
  try {
    let plain = JSON.parse(unescape(__REDUX_STATE__))
    _.each(plain, (val, key)=> {
      reduxState[key] = val;
    })
  } catch (e) {
  }
}

const store = configureStore(reduxState)
store.runSaga(rootSaga)

ReactDOM.render((
// hydrate((
  <Provider store={store}>
    { createRoutes(browserHistory) }
  </Provider>
), document.getElementById('root'))

```

**Comment by probablyup** [comment]
@Nemsae out of curiosity, are you using webpack 4 with `sideEffects: false`? `injectGlobal` is a side effect, so I'm wondering if it's getting pruned out of the bundle.

**Comment by Nemsae** [comment]
@probablyup Interesting, the project is using `"webpack": "^2.2.0",` and I can't find an equivalent option that can be set for that.

**Comment by probablyup** [comment]
Okay that isn’t it then. This is really weird. Any chance you could post a
minimal reproduction repo?

**Comment by iwarner** [comment]
With NextJS and Static deployment I get FOUC also: https://staging.tailwise.com/
Just posting this in case it helps inspect anything that should not rely on the SSR setup.

**Comment by probablyup** [comment]
@iwarner can you open a separate ticket with the environment info and reproduction case? Unless you have the exact same setup as the OP I probably can’t help you without more information

**Comment by iwarner** [comment]
> @iwarner can you open a separate ticket with the environment info and reproduction case? Unless you have the exact same setup as the OP I probably can’t help you without more information

Yeah no worries - it was more just a real life example that you may be able to cream some info from.

I will try and get round to minifying a shorter example to see if any dependancies or code methods are causing my issue outside next or SC

many thanks appreciate the work

**Comment by tannerlinsley** [comment]
I can verify this is happening in react-static's v6 branch. You can see our plugin here: https://github.com/nozzle/react-static/blob/v6/packages/react-static-plugin-styled-components/src/node.api.js#L7-L19

Component styles are extracted properly, but injectGlobal styles are nowhere to be found 😢

**Comment by mxstbr** [comment]
> Component styles are extracted properly, but injectGlobal styles are nowhere to be found 😢

Can you try `styled-components@beta` with the new `createGlobalStyle` API?

**Comment by tannerlinsley** [comment]
Yes I'll give that a try. Either way I'd still be concerned for users on
existing sites using the non beta api

**Comment by probablyup** [comment]
I'm not concerned. I've used it on several large projects and it works fine, so I'm guessing it's something with your setup.

**Comment by tannerlinsley** [comment]
That link I sent you is the extent of the ssr usage. Are there any quirks
in asynchronous ssr environments I should know about?

**Comment by tannerlinsley** [comment]
I wish it were as simple as replicating Next.js style, but like Gatsby (https://github.com/gatsbyjs/gatsby/blob/master/packages/gatsby-plugin-styled-components/src/gatsby-ssr.js), the `jsx => element` and the `element => renderToHtml) processes are detached (to be pluggable).

Did you happen to test this scenario in a multi-threaded environment? The extraction and rendering are happening on the same thread, so I don't suspect it, but it's worth asking.

**Comment by probablyup** [comment]
Given that JS is single threaded, no testing has been done on it, no.

**Comment by tannerlinsley** [comment]
I think you're assuming I was referring to parallelism, of which JS threads are not capable. However, concurrency is commonplace in Node (despite its poor tooling 😜). But I digress.

On a more related note, I do believe I have found the solution to my problem.

TL;DR; - Don't use two instances of `styled-components` (usually inadvertently done via webpack bundling)

It's easy to forget that `injectGlobal` by design cannot utilize context in react (which I'm sure was one of the motivations for the `createGlobalStyle` factory). Because of this, whenever you import/require a module that uses `injectGlobal`, it's safe to assume it is executing immediately most of the time and thus attempting to register the styles with internally imported `Stylesheet` model. This becomes immediately problematic when working with multiple instances of `styled-components` being used throughout a static export process.

Normally styled-component's internal Stylesheet import (`import Stylesheet from 'models/Stylesheet') would return the same model that the `ServerStyleSheet` and `StyleSheetManager` require as well. But in the case of React-Static and many other multi-step-export webpack configurations, it's possible to bundle your application with styled-components as a bundled dependency, and then attempt to extract styles using a different non-bundled instance of styled-component (eg. using a plugin or external API that uses styled-components).

If that's the case (like it was for most react-static, the simple fix is to flag `styled-components` as an webpack external:

```javascript
const webpackConfig = {
  externals: ['styled-components']
}
```

Doing this will ensure that webpack does NOT bundle the `styled-components` dependency, but instead use a `require('styled-components')` during runtime. This will ensure that your `injectGlobal` calls and ServerStylesheet compoennts are coming from the same required instance of `style-components`.

**Comment by probablyup** [comment]
Ah yeah, that'd do it. The issue is somewhat mitigated with the v4 beta by the way since `createGlobalStyle` does consume context.

**Comment by probablyup** [comment]
Now that the TS typings are merged for v4 we'll be releasing the final version soon.

**Comment by tannerlinsley** [comment]
Yeah, I understand most of the team has mentally (and possibly literally) checked out of v3 and are focusing on moving forward with the better designed API in v4. With a project like React-Static, however, we don't have that luxury when hundreds of our users sites use styled-components v3.

Good luck on v4!

**Comment by probablyup** [comment]
@franky47 would you mind actually making a new issue for this so your reproduction doesn't get buried?

**Comment by kitten** [comment]
@franky47 I think your issue is actually a next.js usage issue. Afaik the rendered elements in `_document.js` generate static markup, which might cause issues, since that's not meant to be concurrency safe and it'll also mess with our context-driven global styles.

If you instead include the global style component in `_app.js` it should work correctly: https://github.com/zeit/next.js/#custom-app

**Comment by mxstbr** [comment]
@kitten the official example also uses `_document.js` though: https://github.com/zeit/next.js/blob/canary/examples/with-styled-components/pages/_document.js /cc @timneutkens

**Comment by probablyup** [comment]
Uses it yeah but global styles can't go in there or they won't be present client-side. AFAIK `_document` is only stuff that is in the SSR payload

**Comment by mxstbr** [comment]
Oh yeah I just double checked, I also put it in `_app.js` in my website, which works: https://github.com/mxstbr/mxstbr.com/blob/master/pages/_app.js You're right!

**Comment by leebenson** [comment]
I seem to be having the same issue when trying to bump [ReactQL](https://github.com/leebenson/reactql) to SC v4.

[Here's how SSR is currently wired up](https://github.com/leebenson/reactql/blob/master/src/entry/server.tsx#L48-L105) (with SC v3). Pertinent lines highlighted.

Note the call to Apollo's `getDataFromTree()`.

In SC v4, I _do_ get a render of global styles (via `createGlobalStyle`) but I _don't_ get any styles that would applied to async data, even though my call to `sheet.getStyleElement()` is happening _after_ the call to `getDataFromTree()` and thus should be considered part of the initial style dump.

Logging out the `sheet` object seems to hold a clue:

<img width="548" alt="screenshot 2018-10-17 at 16 05 53" src="https://user-images.githubusercontent.com/719722/47096412-0c7ee380-d227-11e8-9feb-9b5a1515f67a.png">

At first glance, there seems to be a difference between SC v3 and v4 in determining what is considered 'deferred'.

**Comment by probablyup** [comment]
Ah yes, can you do the stylesheet manager wrapper _after_ running `getDataFromTree`?

**Comment by leebenson** [comment]
@probablyup - moving the `<StyleSheetManager>` to a separate block _after_ `getDataFromTree` yielded the same effect.

The only fix I can think of, whilst also rendering out a separate `<Html>` React component and passing `sheet.getStyleElement()` to it, is to run `renderToString` on _both_ parts, i.e.:

```js
// ...

// Create a new styled-components instance
const sheet = new ServerStyleSheet();

const components = (
  <StyleSheetManager sheet={sheet.instance}>
    <ThemeProvider theme={defaultTheme}>
      <ApolloProvider client={client}>
        <StaticRouter location={ctx.request.url} context={routerContext}>
          <Root />
        </StaticRouter>
      </ApolloProvider>
    </ThemeProvider>
  </StyleSheetManager>
);

// Render the Apollo tree
await getDataFromTree(components);

// ...

// Create response HTML
const html = ReactDOMServer.renderToString(components);

// Pass in the rendered props to a final React wrapper
const reactRender = ReactDOMServer.renderToString(
  <Html
    css={output.client.main("css")!}
    helmet={Helmet.renderStatic()}
    html={html}
    js={output.client.main("js")!}
    styles={sheet.getStyleElement()}
    window={{
      __APOLLO_STATE__: client.extract(),
    }} />,
);

// Set the return type to `text/html`, and stream the response back to
// the client
ctx.type = "text/html";
ctx.body = `<!DOCTYPE html>${reactRender}`;
```

.. which is fine, although I haven't tested the performance cost of running two separate React render chains yet. Hopefully not too much of an impact.

My main point is that this is a material change vs. SC v3. In the previous version, it wasn't necessary to call `renderToString` twice. I could pass in `sheet.getStyleElement()` to the component that got rendered out and it worked fine.

**Comment by probablyup** [comment]
@leebenson are you using react-apollo latest? Supposedly there are some issues with earlier versions and new context etc

**Comment by probablyup** [comment]
Yes try react apollo 2.3+ if it isn’t already installed

**Comment by good-idea** [comment]
Certain styles not working / acting funny can be caused by circular dependencies.
I use `circular-dependency-plugin` when webpacking my app for the frontend - but didn't have anything looking for these warnings on the server, so it's easy to miss!

**Comment by Yzrsah** [comment]
I've seen a bunch of people attempt to get Material-UI + Styled Components working on Next.JS, but nobody ever figured it out.

All of the examples/attempts make FOUC.

Here's 1 attempt:

https://github.com/williamluke4/nextjs-materialui-styledcomponents
(after updating deps, add "webpack": "^4.29.4" to fix build error)

This produces FOUC every time.

**Comment by elado** [comment]
I had the same issue (with v4).
Turned out I was collecting styles before some store hydration happened, which changed the react tree. I moved the collection after the hydration, and no longer having the first render of unstyled components.

**Comment by kitten** [comment]
@elado you could try out v5 if you don't mind where this shouldn't be an issue anymore 🤔

I'll close this issue for now as all SSR ordering and rendering issues should be solved in v5, which will be released soon, or have been unrelated to the underlying issues.

**Comment by nghiepit** [comment]
@elado While waiting for v5 release, you can play with the trick, here an example for `Next.js`

```jsx
import React from "react";
import App, { Container } from "next/app";

class MyApp extends App {
  state = {
    ready: false
  };

  componentDidMount() {
    this.setState({
      ready: true
    });
  }

  render() {
    const { Component, pageProps } = this.props;

    return (
      <Container>
        {!this.state.ready && <div className="fouc" />}
        <Component {...pageProps} />
        <style jsx global>{`
          .fouc + * {
            visibility: hidden !important;
          }
        `}</style>
      </Container>
    );
  }
}

export default MyApp;

```

**Comment by bitworking** [comment]
> I had the same issue (with v4).
> Turned out I was collecting styles before some store hydration happened, which changed the react tree. I moved the collection after the hydration, and no longer having the first render of unstyled components.

@elado Can you please explain a little bit more how you did this? I'm using nextjs with Apollo Client (getDataFromTree) and I have the problem that sometimes some styles are missing on SSR. It sounds like this could be the problem.
